File: backend/app/shared/validators.py
import re

def is_valid_rut(rut: str) -> bool:
    """
    Valida un RUT chileno usando el algoritmo de Módulo 11.
    Acepta formatos: '12.345.678-9', '12345678-9', '123456789'
    El cuerpo puede tener de 6 a 8 dígitos (RUT van desde 1.000.000 hasta 99.999.999).
    """
    if not rut:
        return False

    # Eliminar puntos, guiones y espacios; convertir a mayúsculas
    rut_clean = rut.replace(".", "").replace("-", "").replace(" ", "").upper()

    # El cuerpo puede tener 6, 7 u 8 dígitos + 1 dígito verificador (0-9 o K)
    # Total: 7, 8 o 9 caracteres
    if not re.match(r"^\d{6,8}[0-9K]$", rut_clean):
        return False

    body = rut_clean[:-1]
    dv = rut_clean[-1]

    # Algoritmo Módulo 11
    suma = 0
    multiplo = 2
    for digit in reversed(body):
        suma += int(digit) * multiplo
        multiplo = 2 if multiplo == 7 else multiplo + 1

    remainder = suma % 11
    expected_dv_value = 11 - remainder

    if expected_dv_value == 11:
        expected_dv = "0"
    elif expected_dv_value == 10:
        expected_dv = "K"
    else:
        expected_dv = str(expected_dv_value)

    return dv == expected_dv

File: frontend/front-descanso-premium/src/utils/validators.ts
export const validateRut = (rut: string): boolean => {
  if (!rut) return false;
  
  // Limpiar el RUT de puntos, guiones y espacios
  const cleanRut = rut.replace(/[^0-9kK]/g, '').toUpperCase();
  
  // Un RUT chileno tiene mínimo 7 caracteres (6 dígitos de cuerpo + 1 DV)
  // y máximo 9 (8 dígitos de cuerpo + 1 DV)
  if (cleanRut.length < 7 || cleanRut.length > 9) return false;
  
  const body = cleanRut.slice(0, -1);
  const dv = cleanRut.slice(-1);
  
  // Calcular dígito verificador con algoritmo Módulo 11
  let sum = 0;
  let multiplier = 2;
  
  for (let i = body.length - 1; i >= 0; i--) {
    sum += parseInt(body[i]) * multiplier;
    multiplier = multiplier === 7 ? 2 : multiplier + 1;
  }
  
  const remainder = sum % 11;
  const expectedDv = 11 - remainder;
  
  let expectedDvStr: string;
  if (expectedDv === 11) expectedDvStr = '0';
  else if (expectedDv === 10) expectedDvStr = 'K';
  else expectedDvStr = expectedDv.toString();
  
  return dv === expectedDvStr;
};

export const formatRut = (rut: string): string => {
  // Eliminar todo lo que no sea número o K
  let cleanRut = rut.replace(/[^0-9kK]/g, '').toUpperCase();
  
  if (cleanRut.length === 0) return '';
  if (cleanRut.length <= 1) return cleanRut;
  
  // Extraer cuerpo y dígito verificador
  const dv = cleanRut.slice(-1);
  const body = cleanRut.slice(0, -1);
  
  // Formatear cuerpo con puntos
  const formattedBody = body.replace(/\B(?=(\d{3})+(?!\d))/g, '.');
  
  return `${formattedBody}-${dv}`;
};
